### PR TITLE
Remove top level version

### PR DESCRIPTION
**Pull request by natalieparellano**
Signed-off-by: Natalie Arellano <narellano@vmware.com>

### CODE REVIEW

**Comment by ekcasey** [comment]
Love how organized this is :)

File: builder.go [hunk 1]
```diff
@@ -165,6 +165,12 @@ func (p BuildPlan) find(bp Buildpack) buildpackPlan {
 			}
 		}
 	}
+	if bp.API == "0.2" {
+		for i, entry := range out {
+			entry.convertMetadataToVersion()
+			out[i] = entry
+		}
+	}
```
Comment (natalieparellano) [inline builder.go:173]: Done

File: detector.go [hunk 2]
```diff
@@ -47,6 +47,29 @@ type Require struct {
...
+	}
+	return false
+}
+
+func (r *Require) convertMetadataToVersion() {
+	if version, ok := r.Metadata["version"]; ok {
+		r.Version = fmt.Sprintf("%v", version)
+	}
+}
+
```
Comment (natalieparellano) [inline detector.go:62]: Done

File: detector.go [hunk 3]
```diff
@@ -251,6 +274,19 @@ func (bp *BuildpackTOML) detect(c *DetectConfig) detectRun {
 	if _, err := toml.DecodeFile(planPath, &t); err != nil {
 		return detectRun{Code: -1, Err: err}
 	}
+	if bp.API == "0.2" {
+		if t.hasInconsistentVersions() || t.Or.hasInconsistentVersions() {
+			// fail
+		}
+	}
```
Comment (natalieparellano) [inline detector.go:281]: Done

File: detector.go [hunk 4]
```diff
@@ -361,17 +411,53 @@ func hasID(bps []Buildpack, id string) bool {
...
 type detectRun struct {
 	planSections
-	Or     []planSections `toml:"or"`
-	Output []byte         `toml:"-"`
-	Code   int            `toml:"-"`
-	Err    error          `toml:"-"`
+	Or     planSectionsList `toml:"or"`
+	Output []byte           `toml:"-"`
+	Code   int              `toml:"-"`
+	Err    error            `toml:"-"`
```
Comment (natalieparellano) [inline detector.go:417]: Done

File: detector.go [hunk 5]
```diff
@@ -361,17 +411,53 @@ func hasID(bps []Buildpack, id string) bool {
...
 
+func (p *planSections) hasInconsistentVersions() bool {
+	for _, req := range p.Requires {
+		if req.hasInconsistentVersions() {
+			return true
+		}
+	}
+	return false
+}
+
```
Comment (natalieparellano) [inline detector.go:433]: Done

File: detector.go [hunk 6]
```diff
@@ -361,17 +411,53 @@ func hasID(bps []Buildpack, id string) bool {
...
+func (p *planSectionsList) hasInconsistentVersions() bool {
+	for _, planSection := range *p {
+		if planSection.hasInconsistentVersions() {
+			return true
+		}
+	}
+	return false
+}
+
```
Comment (natalieparellano) [inline detector.go:452]: Done

File: detector.go [hunk 7]
```diff
@@ -47,6 +47,29 @@ type Require struct {
...
+func (r *Require) convertVersionToMetadata() {
+	if r.Version != "" {
+		if r.Metadata == nil {
+			r.Metadata = make(map[string]interface{})
+		}
+		r.Metadata["version"] = r.Version
+		r.Version = ""
+	}
+}
+
```
Comment (natalieparellano) [inline detector.go:72]: Done

File: detector.go [hunk 8]
```diff
@@ -269,6 +305,13 @@ func (bg BuildpackGroup) Detect(c *DetectConfig) (BuildpackGroup, BuildPlan, err
 	} else if err == errFailedDetection {
 		err = NewLifecycleError(err, ErrTypeFailedDetection)
 	}
+	for i, entry := range entries {
+		for j, req := range entry.Requires {
+			req.convertVersionToMetadata()
+			entry.Requires[j] = req
+		}
+		entries[i] = entry
+	}
```
Comment (natalieparellano) [inline detector.go:314]: Done

File: detector.go [hunk 9]
```diff
@@ -322,6 +365,13 @@ func (bo BuildpackOrder) Detect(c *DetectConfig) (BuildpackGroup, BuildPlan, err
...
```
Comment (natalieparellano) [inline detector.go:374]: Done

**Comment by natalieparellano** [comment]
Closing this draft PR, will open a new one for the "real" code.

**Pull request closed without merging**